fast_carpenter/masked_tree.py

```diff
@@ -118,3 +127,137 @@ def mask_df(df, mask, start_event):
     else:
         masked = df.loc[mask]
     return masked
+
+
+class MaskedTrees(object):
+    nonOverloadedFunctions = [
+        'pandas', 'array', 'arrays', 'mask', 'unmasked_array',
+        'unmasked_arrays', 'apply_mask', 'trees', '_mask', 'event_ranger',
+        '_index', '__getitem__', 'reset_mask', 'new_variable', 'keys',
+    ]
+
+    def __init__(self, trees, event_ranger, mask=None):
+        self.pandas = types.SimpleNamespace(df=self._df)
+        self.trees = {}
+        self.provenance = {}
+        self._mask = mask
+        # TODO: need to make sure event_ranger act very low level to prevent reads.
+        # currently, that is broken and only works for DFs
+        # MAskedTrees is probably too late to deal with event ranges
+        self.event_ranger = event_ranger
+        self._index = {}
+        for name, tree in trees.items():
+            self.trees[name] = MaskedUprootTree(tree, event_ranger, mask)
+            self.provenance[name] = name.split('/') if '/' in name else [name]
+            self._index.update(recursive_index(self._index, self.provenance[name], self.trees[name]))
+        self._index = create_aliases(self._index)
+
+    def __getitem__(self, key):
+        return self._index[key]
+
+    def _df(self, *args, **kwargs):
+        return self.arrays(*args, **kwargs, outputtype=pd.DataFrame)
+
+    def unmasked_array(self, *args, **kwargs):
+        return self.tree.array(*args, **kwargs)
+
+    def unmasked_arrays(self, *args, **kwargs):
+        return self.tree.arrays(*args, **kwargs)
+
+    def array(self, *args, **kwargs):
+        array, exception = None, None
+        try:
+            value = self._index[args[0]]
+            array = value.array(*args[1:], **kwargs)
```

Comment:
I'm trying to understand this implementation.  Is the functionality here to give people the flexibility to get the same result calling this function in several different ways? e.g.
```
chunk.tree.array("treename", "variable.subvar")
chunk.tree.array("treename.variable.subvar")
chunk.tree.array("treename__DOT__variable__DOT__subvar")
```
and expect the same returned object?  

I might be wrong there, but if the answer is yes to the above then:
1. If you get a numpy returned from line 170, will that object have the `.array()` method itself?
2. I think it might make sense to provide a different function for each of these functionalities.  the version with `__DOT__` in the name is essentially an internal API - I don't think we should expect users to make use of it directly in their own custom stages.

Reply:
Yes.

1. If a numpy object is returned, it will have no array method &rarr; an exception will be raise.

2. Yes, it is in the spec (see `test_multitree_aliases`). This was a workaround for internal usage AFAIR.

Noticed that that block
```python
        array = None
        try:
            ...
        except Exception as e:
            exception = e
        if array is None:
            raise exception

```
is redundant, so removed it.